### DISCUSSION TITLE
Metal with Zig

### DISCUSSION THREAD

**Discussion by cztomsik** [ggml-org/llama.cpp#2264] [General]
I'm not sure where to ask because it's not strictly an issue with llama.cpp but when I try to use llama.cpp with Zig (with metal enabled), I get a really weird bug in runtime, which looks like this:

<img width="818" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/3526922/d980f6d2-9678-4ef0-9539-06cba54b080e">

When I compile llama.cpp using make, it works fine, it only fails with Zig, and what's especially interesting, it fails during runtime, when the metal shaders are compiled. I've checked that the metal file is the same, even the llama.cpp codebase is the same, everything is the same, except the compiler.

So I was thinking okay, maybe apple clang links something in a different way but even the `otool` and `DYLD_PRINT_LIBRARIES` showed exactly the same thing.

Any ideas would be welcome, thanks.

**Comment by amusingimpala75** [comment]
Did you ever figure out how to fix this? I'm considering checking to see if just having the build.zig explicitly call the Makefile and then link it in, rather than using the builtin C/C++ compilation, works.

**Reply by cztomsik** [reply]
You will not believe it but the problem (my problem, I'm not sure if anyone else was experiencing this particular thing, I couldn't find a single mention anywhere) was actually caused by brew... I did `brew update` and `brew upgrade` and suddenly metal started working fine. However, I was not using the build.zig from this repo (it does not include metal), but my own https://github.com/cztomsik/dan/blob/bf01f31a9819c53ac05b043a2726402319ba611d/build.zig

BTW: the crazy thing is that it was working fine with apple clang, the problem was only when compiled with zig.
Also, I've found that it's not just when used from build.zig, but also, if you use zig as cross-compiler (zig cc), really weird...

**Reply by amusingimpala75** [reply]
Yeah same. GCC worked fine, linking it using exe.addCSourceFile didn’t. When did you `brew upgrade`? Zig 0.11.0 just released yesterday, maybe that’s what fixed it.

**Reply by cztomsik** [reply]
~week ago
